src/Servers/Kestrel/test/InMemory.FunctionalTests/RequestTests.cs

```diff
@@ -727,6 +727,129 @@ public async Task ContentLengthReadAsyncPipeReader()
             }
         }
 
+        [Fact]
+        public async Task ContentLengthReadAsyncPipeReaderBufferRequestBody()
+        {
+            var testContext = new TestServiceContext(LoggerFactory);
+
+            using (var server = new TestServer(async httpContext =>
+            {
+                var readResult = await httpContext.Request.BodyReader.ReadAsync();
+                // This will hang if 0 content length is not assumed by the server
+                Assert.Equal(5, readResult.Buffer.Length);
+                httpContext.Request.BodyReader.AdvanceTo(readResult.Buffer.Start, readResult.Buffer.End);
+                readResult = await httpContext.Request.BodyReader.ReadAsync();
+                Assert.Equal(5, readResult.Buffer.Length);
```

Comment:
Do you have any tests that attempt more than one read *after* getting a completed read result?